### ISSUE TITLE
hotlink protection for hls

### ISSUE THREAD

**Issue by myrmidon-media** [sergey-dryabzhinsky/nginx-rtmp-module#56]
Would be great to have some sort of hotlink protection for hls links, so a m3u8 link is only valid for a certain period of time, similar to what wowza or nimble streamer have.

**Comment by Tralapo** [comment]
Nginx-rtmp does support secure links, but I never got that working. Maybe you can: https://rarut.wordpress.com/2013/06/09/secure-links-in-nginx-rtmp/

In my case I only had to prevent streaming outside my own website, so I went with valid-referes. Does the job good enough for me http://nginx.org/en/docs/http/ngx_http_referer_module.html

**Comment by misiek08** [comment]
You can write script to generate secure links and secure it via nginx with
secure links module or auth_request directive and second script. It won't
protect you from downloading your videos, but people will be able to watch
only from website that know your secret key.

2015-12-12 16:27 GMT+01:00 Tralapo <notifications@github.com>:

**Comment by myrmidon-media** [comment]
Yes but with secure link module if we put an small expiration time for example 5 mins. The m3u8 link will become unavailable after this 5 mins and the player needs to reload the playlists from time to time for the playback to work.

**Comment by misiek08** [comment]
You need to generate m3u8 dynamically by script and do not secure m3u8 but secure links to fragments inside.

**Comment by myrmidon-media** [comment]
How are you going to secure the fragments if the fragments are generated by nginx?

**Comment by misiek08** [comment]
You have original m3u8 and you can parse it and replace content for every user. You can use hooks too of course.

**Comment by sergey-dryabzhinsky** [comment]
There is no currently any event handler for playlist update. So you can't catch that moment and "fix" playlist.
I wrote some code already that do work. It will be pushed here soon.
Stay tuned.

**Comment by sergey-dryabzhinsky** [comment]
You can check code from branch ```notify-on-playlist-hls-dash```. Look in #58

**Comment by matiaspl** [comment]
As a temporary measure it can be done though filesystem inotify callbacks.

Matt

2015-12-17 1:41 GMT+01:00 Sergey <notifications@github.com>:

**Comment by AlexRevin** [comment]
thanks @sergey-dryabzhinsky. Could you please drop a line about usage as it's not obvious for many from the source code.

**Comment by sergey-dryabzhinsky** [comment]
Done: https://github.com/sergey-dryabzhinsky/nginx-rtmp-module/wiki/New-Directives-%28fork%29

**Comment by myrmidon-media** [comment]
So anybody has a working example of how can ```notify-on-playlist-hls-dash``` can be used for hotlink protection?

Thanks

**Comment by benwilber** [comment]
I wrote a blog on how I did this http://benwilber.github.io/streamboat.tv/nginx/varnish/hotlink/video/streaming/2016/02/06/streamboat-tv-hotlink-protection-with-nginx-and-varnish.html

**Comment by AlexRevin** [comment]
@benwilber HLS protection is a very tricky thing and there's no guaranteed method to bypass it. Even AES key encryption gets hacked.

What I don't like in your approach:

1. Customers on 3G/4G mobile devices moving from tower to tower will get disconnected as IP will get changed

2. it only takes 1 request to your site to get a token key which can be re-distributed to poachers viewer.

I would rather suggest generating custom m3u8 manifest urls and rate limiting with ```ngx_http_limit_req_module``` over the url key, so that each key can be used only once per 4-5 seconds.

The paranoid mode which will work in all cases despite re-distributing the stream with own CDN is a one-time-token for .ts segment.

In brief:

1. read m3u8 on callback and cache it
2. generate dynamically m3u8 manifest for one specific user

- [ ]   generate and store a unique segment key in redis
- [ ]   apply the segment key to .ts segment url

3. In nginx on .ts request you:

- [ ]   read the key from redis via the [redis2-nginx-module](https://github.com/openresty/redis2-nginx-module) 
- [ ]   if key is present - all good, if not - 404
- [ ]   delete the key from redis

In this case you will have exactly **1-time** readable segment(or you can tune it for 2-3 times) which cannot be re-distributed to multiple users.

This strategy will require some effort and high-available app server, but so far combining those 2 approaches is even more secure than AES encryption as tls keys are passed to client side and can be copied.

**Comment by sergey-dryabzhinsky** [comment]
@AlexRevin there is maybe another way:

- Add option to `hls` module to define `BASE URI` or `BASE URL` for m3u segments and `.ts` segments
- Set it to dynamic site / cms. Which can route it and do internal request for `.m3u8` and `.ts` files from some hidden `nginx` location.
- That site / cms can check access rights by tokens / cookies / etc.
- Or it may be another nginx location with fast `lua` / `perl` or `redis` code.
